### PR TITLE
[NU-1092] fixed ExpressionSuggester

### PR DESCRIPTION
**Pull request by JulianWielga**
## Checklist before merge
- [ ] Related issue ID is placed at the beginning of PR title in \[brackets\] (can be GH issue or Nu Jira issue)
- [ ] Code is cleaned from temporary changes and commented out lines
- [ ] Parts of the code that are not easy to understand are documented in the code
- [ ] Changes are covered by automated tests
- [ ] Showcase in dev-application.conf added to demonstrate the feature
- [ ] Documentation added or updated
- [ ] Added entry in _Changelog.md_ describing the change from the perspective of a public distribution user
- [ ] Added _MigrationGuide.md_ entry in the appropriate subcategory if introducing a breaking change
- [ ] Verify that PR will be squashed during merge

### CODE REVIEW

File: designer/client/src/components/graph/node-modal/editors/expression/CustomAceEditorCompleter.tsx [hunk 1]
```diff
@@ -30,11 +30,21 @@ function isSpelTokenAllowed(iterator, modeId): boolean {
...
-        activated: boolean;
-        openPopup: (editor: Editor, prefix: string, keepPopupPosition: boolean) => void;
-    };
+declare module "ace-builds" {
+    // eslint-disable-next-line @typescript-eslint/no-namespace
+    namespace Ace {
+        interface Autocomplete {
+            base: Ace.Point;
+            editor: Ace.Editor;
+            activated: boolean;
```
Comment (Dzuming) [inline designer/client/src/components/graph/node-modal/editors/expression/CustomAceEditorCompleter.tsx:39]: Do we need it? looks like it's not called and if i understand correctly, we extend predefined Ace.Autocomplete interface here

File: designer/client/src/components/graph/node-modal/editors/expression/CustomAceEditorCompleter.tsx [hunk 2]
```diff
@@ -30,11 +30,21 @@ function isSpelTokenAllowed(iterator, modeId): boolean {
...
+            openPopup: (this: Autocomplete, editor: Ace.Editor, prefix: string, keepPopupPosition: boolean) => void;
```
Comment (Dzuming) [inline designer/client/src/components/graph/node-modal/editors/expression/CustomAceEditorCompleter.tsx:40]: Do we need it? looks like it's not called and if i understand correctly, we extend predefined Ace.Autocomplete interface here

File: designer/client/src/components/graph/node-modal/editors/expression/ExpressionSuggester.ts [hunk 3]
```diff
@@ -2,37 +2,127 @@ import HttpService from "../../../../../http/HttpService";
...
+        return () => this.off(eventName, callback);
+    }
+
+    off<E extends SuggesterEvent>(eventName: E | `${E}`, callback: EventsMap[E]): void {
+        this.getHandlers(eventName).delete(callback);
+    }
+
+    suggestionsFor(expression: string, caretPosition2d: CaretPosition2d): Promise<ExpressionSuggestion[]> {
+        this.isLoading = true;
+        const currentRequest = this.httpService
```
[the remark this answers is not in the document]
Comment (JulianWielga) [reply designer/client/src/components/graph/node-modal/editors/expression/ExpressionSuggester.ts:89]: it was to have a clearer situation few lines later:

        return currentRequest
            .then(() => this.latestRequest)

also wanted to have this assignment as clear as possible, not multiline promise.  

        this.latestRequest = currentRequest;

File: designer/client/src/components/graph/node-modal/editors/expression/ExpressionSuggester.ts [hunk 4]
```diff
@@ -2,37 +2,127 @@ import HttpService from "../../../../../http/HttpService";
...
 
 export interface ExpressionSuggester {
     suggestionsFor(inputValue: string, caretPosition2d: CaretPosition2d): Promise<ExpressionSuggestion[]>;
+
+    cancel(): void;
+}
+
+export enum SuggesterEvent {
+    stateChange = "stateChange",
+    stateChange2 = "stateChange2",
```
[the remark this answers is not in the document]
Comment (JulianWielga) [reply designer/client/src/components/graph/node-modal/editors/expression/ExpressionSuggester.ts:35]: for sure - not ;)

File: designer/client/src/components/graph/node-modal/editors/expression/CustomAceEditorCompleter.tsx (hunk 1 above)
Comment (JulianWielga) [reply designer/client/src/components/graph/node-modal/editors/expression/CustomAceEditorCompleter.tsx:39]: this is not an extension but more supplement to what was delivered with holes. maybe it's not used by us anymore, maybe I used it while debugging, but it's identified already and could help in future

File: designer/client/src/components/graph/node-modal/editors/expression/CustomAceEditorCompleter.tsx (hunk 2 above)
Comment (JulianWielga) [reply designer/client/src/components/graph/node-modal/editors/expression/CustomAceEditorCompleter.tsx:40]: ^

File: designer/client/src/components/graph/node-modal/editors/expression/CustomAceEditorCompleter.tsx (hunk 1 above)
Comment (Dzuming) [reply designer/client/src/components/graph/node-modal/editors/expression/CustomAceEditorCompleter.tsx:39]: Ok, I thought there was our default propose which we use to extend the Autocomplete interface, but it it's existing Autocomplete props, that's ok.

**Review by Dzuming — APPROVED** [review]

**Pull request merged**